### PR TITLE
Add Pytorch Memory APIs

### LINKED ISSUE #853: Initial pytorch memory APIs for Spyre
## Issue Summary

Add memory tracking and management APIs to `torch.spyre` to expose memory allocation metrics and allow resetting memory statistics. These APIs will enable users to monitor and debug memory usage in SPYRE-based operations.

### 1. `torch.spyre.memory_allocated()`

**Purpose:** Get the current memory allocated by SPYRE operations.

**Signature:**
```python
def memory_allocated() -> int:
    """
    Returns the current memory allocated by SPYRE in bytes.

    Returns:
        int: Current memory allocated in bytes

    Example:
        >>> import torch
        >>> torch.spyre.memory_allocated()
        1024000  # 1MB in bytes
    """
```

**Expected Behavior:**
- Returns the total memory currently allocated by all active SPYRE operations
- Returns an integer value in bytes
- Should reflect real-time memory usage

---

### 2. `torch.spyre.max_memory_allocated()`

**Purpose:** Get the peak memory allocated by SPYRE operations since the last reset.

**Signature:**
```python
def max_memory_allocated() -> int:
    """
    Returns the maximum memory that has been allocated by SPYRE since
    the last reset of peak memory statistics.

    Returns:
        int: Peak memory allocated in bytes

    Example:
        >>> import torch
        >>> torch.spyre.max_memory_allocated()
        5242880  # 5MB in bytes
    """
```

**Expected Behavior:**
- Tracks the maximum memory allocated since initialization or last reset
- Returns an integer value in bytes
- Persists across multiple SPYRE operations until explicitly reset

---

### 3. `torch.spyre.reset_peak_memory_stats()`

**Purpose:** Reset the peak memory statistics counter.

**Signature:**
```python
def reset_peak_memory_stats() -> None:
    """
    Reset the peak memory statistics for SPYRE operations.

    After calling this function, max_memory_allocated() will return the
    current memory usage until a larger allocation occurs.

    Example:
        >>> import torch
        >>> torch.spyre.max_memory_allocated()
        5242880
        >>> torch.spyre.reset_peak_memory_stats()
        >>> torch.spyre.max_memory_allocated()
        1024000  # Now reflects current allocation after reset
    """
```

**Expected Behavior:**
- Resets the peak memory counter
- Current memory usage becomes the new baseline for peak tracking
- Does not deallocate any memory, only resets statistics

---

### 4. `torch.spyre.reset_accumulated_memory_stats()`

**Purpose:** Reset accumulated memory statistics across all SPYRE operations.

**Signature:**
```python
def reset_accumulated_memory_stats() -> None:
    """
    Reset all accumulated memory statistics for SPYRE operations.

    This resets both current and peak memory tracking counters, providing
    a clean slate for memory profiling in SPYRE operations.

    Example:
        >>> import torch
        >>> torch.spyre.memory_allocated()
        1024000
        >>> torch.spyre.max_memory_allocated()
        5242880
        >>> torch.spyre.reset_accumulated_memory_stats()
        >>> torch.spyre.memory_allocated()
        0
        >>> torch.spyre.max_memory_allocated()
        0
    """
```

**Expected Behavior:**
- Resets both current memory and peak memory counters
- Sets all memory statistics to zero
- Does not affect actual memory allocations, only the tracking counters
- Useful for profiling specific code sections in isolation

---

### Design Pattern
- Follow the same pattern as `torch.cuda.memory_allocated()` and `torch.cuda.max_memory_allocated()` for consistency
- Maintain backward compatibility with existing SPYRE API

### Memory Tracking
- Implement a thread-safe memory tracker for SPYRE allocations
- Track allocations and deallocations
- Maintain peak memory watermark

### Integration Points
- Hook into SPYRE allocation/deallocation routines
- Register memory tracking in the SPYRE initialization
- Ensure minimal performance overhead

### Testing Requirements
1. Unit tests for each API function
2. Test concurrent SPYRE operations with memory tracking
3. Test reset functionality
4. Test edge cases (no allocations, rapid allocations/deallocations)

### Documentation Requirements
1. Add docstrings to all API functions
2. Update PyTorch documentation
3. Add examples to SPYRE tutorial/guide
4. Add entry to release notes

## Acceptance Criteria

- [ ] `torch.spyre.memory_allocated()` returns current SPYRE memory in bytes
- [ ] `torch.spyre.max_memory_allocated()` returns peak SPYRE memory in bytes
- [ ] `torch.spyre.reset_peak_memory_stats()` resets peak counter only
- [ ] `torch.spyre.reset_accumulated_memory_stats()` resets all counters
- [ ] All APIs are thread-safe
- [ ] All APIs follow PyTorch naming conventions
- [ ] Comprehensive unit tests with >90% code coverage
- [ ] Performance impact is negligible (<1% overhead)
- [ ] Documentation is complete and clear
- [ ] API behavior matches CUDA memory APIs for consistency

### LINKED ISSUE #941: Configure profiler to capture memory statistics when enabled
Depends on: #853 

## Description
Add functionality to capture and report memory statistics during profiling sessions using PyTorch's profiler API with privateuse1 backend support. When the `profiler_memory=true` flag is enabled, the profiler collects detailed memory usage metrics.

## Requirements
- [x] Track memory allocations over time
- [x] Record peak memory usage
- [x] Output memory report in a structured format
- [x] Support privateuse1 backend device allocation tracking
- [x] Integrate with torch.profiler API

## Configuration
Set the following environment variable or configuration flag to enable:
```
profiler_memory=true
PYTORCH_PROFILER_BACKEND=privateuse1
```

## Implementation Details
PyTorch profiler integration with privateuse1 backend:
1. Initialize memory tracking using `torch.profiler.profile()`
2. Enable memory history with `record_shapes=True`, `profile_memory=True`
3. Implement reportMemoryUsageToProfiler in the runtime where capturing memory allocations/deallocation 
4. Record allocation patterns from device allocations
5. Generate memory profile report on completion using profiler.key_averages()
6. Visualize using perfetto and tensor board

## Output Format
Memory statistics captured in structured format including:
- Total device memory used (privateuse1 backend)
- Peak memory consumption per operation
- Memory allocated by operation/kernel
- Timeline of memory usage changes via profiler traces

## Expected Behavior
When enabled, the profiler:
1. Initializes memory tracking at context manager entry
2. Captures all device allocations/deallocations for privateuse1
3. Records memory states at operation boundaries
4. Generates comprehensive profile report on context exit
5. Outputs memory timeline and peak usage statistics

## Related Documentation
- PyTorch Profiler: https://pytorch.org/docs/stable/profiler.html
- PrivateUse1 Backend Support: Custom device backend integration
- Memory Profiling: torch.profiler memory event tracking

### PR DESCRIPTION
**Pull request by SilverSoldier**
#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Adds Pytorch CUDA-like APIs tracking current and peak memory allocation along with resets.

#### Which issue(s) this PR is related to:
Fixes #853 
Fixes #941 
#### Special notes for your reviewer:

Quick test:
```
import torch
import torch_spyre
x = torch.ones((64, 64), dtype=torch.float16, device='spyre')
torch.spyre.memory.memory_allocated()
```

#### Does this PR introduce a user-facing change?

Introduces new APIs:
```
torch.spyre.memory.memory_allocated()
torch.spyre.memory.max_memory_allocated()
torch.spyre.memory.reset_peak_memory_stats()
torch.spyre.memory.reset_accumulated_memory_stats()
```

#### Additional note:
If these changes look good, I can other missing relevant APIs such as `memory_summary` etc. 

Also, when VF mode block memory allocation is implemented, we can use the pool related counters to manage those memory statistics.

### CODE REVIEW

File: torch_spyre/csrc/spyre_mem.cpp [hunk 1]
```diff
@@ -436,6 +436,8 @@ auto copy_device_to_host(const at::Tensor& self, const at::Tensor& dst) {
 // allocated memory not the actual pointer
 struct SpyreAllocator final : public at::Allocator {
  private:
+  DeviceStats stats;
+  StatTypes stat_types = {true, false, false}; // Setting AGGREGATE to True
```
[the remark this answers is not in the document]
Comment (SilverSoldier) [reply torch_spyre/csrc/spyre_mem.cpp]: Added the meanings of the different types

File: torch_spyre/csrc/spyre_mem.cpp [hunk 2]
```diff
@@ -499,6 +511,34 @@ struct SpyreAllocator final : public at::Allocator {
     // spyre_copy_from(reinterpret_cast<spyre_ptr_t>(dest),
     // reinterpret_cast<spyre_ptr_t>(src));
   }
+
+  void recordAlloc(size_t nbytes) {
+    for_each_selected_stat_type(stat_types, [&](size_t stat_type) {
```
[the remark this answers is not in the document]
Comment (SilverSoldier) [reply torch_spyre/csrc/spyre_mem.cpp]: pytorch `Allocator.h` has the `for_each_selected_stat_type` function

[the remark this answers is not in the document]
Comment (SilverSoldier) [reply torch_spyre/csrc/spyre_mem.cpp]: https://github.com/pytorch/pytorch/blob/915982a48846930ed6b47213b38f77b7f1eba75e/c10/core/Allocator.h#L393

**Review by thoangtrvn — CHANGES_REQUESTED** [review]
Thanks so much. I have a few comments.

**Comment by SilverSoldier** [comment]
@thoangtrvn I have made the changes suggested. Now, we use the python APIs exposed by `torch.accelerator.memory` and rely on it to call the underlying C++ calls.

However, I was not able to completely do away with the functions in our `memory.py` as discussed above. We need to add a guard check for whether SpyreAllocator is initialized or not to avoid Pytorch Runtime Error when trying to find the DeviceAllocator for Spyre.

Please review the changes.

**Comment by kmehant** [comment]
@SilverSoldier Lets create the spyre allocator during import which will solve this and we can avoid the guard code - ref impl - https://github.com/kmehant/torch-spyre/commit/31f373db906a7c16b95ed5dda72048d25eb8a432

**Comment by kaoutar55** [comment]
Thanks for the work here @SilverSoldier — the C++ stat tracking in SpyreAllocator looks solid.

I agree with @kmehant's suggestion to eagerly initialize the allocator at import time. It eliminates the need for the _allocator_ready() guard and the entire memory.py wrapper file, letting us just re-export torch.accelerator.memory.* directly. That's less code to maintain and keeps us aligned with how other PyTorch backends handle this.

A couple of things I noticed while reviewing the current diff:

resetAccumulatedStats / resetPeakStats appear swapped: resetAccumulatedStats calls reset_peak() and resetPeakStats calls reset_accumulated(). Can you double-check that?

Thread safety on stats_ :recordAlloc and recordRelease mutate the static stats_ member on every allocation/deallocation without any visible locking. If allocations can happen concurrently from different threads, this could race. Worth considering whether a mutex or atomics are needed here.

Test coverage for the edge case:  it would be good to add a test that calls memory_allocated() before any Spyre tensor has been created, since that's the exact scenario driving the guard discussion.

**Comment by SilverSoldier** [comment]
@kmehant Can you confirm that this doesn't throw error when directly `import`ing torch.acclerator without any guards?

This is a sample code which threw the error:
```
import torch
import torch_spyre

x = torch.rand((64, 64), dtype=torch.float16)
print(torch.spyre.memory_allocated())
print(torch.spyre.max_memory_allocated())

x = x.to(device='spyre')
print(torch.spyre.memory_allocated())
print(torch.spyre.max_memory_allocated())
```

I will let @thoangtrvn and others take the call whether it is okay to do this though. Bypassing the lazy initialization might not be expected behaviour, so I'm not sure what/who all this would affect.

**Comment by SilverSoldier** [comment]
Thanks for the review @kaoutar55 
> Thanks for the work here @SilverSoldier — the C++ stat tracking in SpyreAllocator looks solid.
> 
> I agree with @kmehant's suggestion to eagerly initialize the allocator at import time. It eliminates the need for the _allocator_ready() guard and the entire memory.py wrapper file, letting us just re-export torch.accelerator.memory.* directly. That's less code to maintain and keeps us aligned with how other PyTorch backends handle this.

That is true, but the current code has some order for initialization, for instance starting the flex runtime, ensuring initialization is called exactly once with a flag etc. Initializing the allocator out of order, before the runtime is initialized, could cause problems. I definitely want some confirmation from the runtime team that this won't cause problems before we go ahead with this. 

> 
> A couple of things I noticed while reviewing the current diff:
> 
> resetAccumulatedStats / resetPeakStats appear swapped: resetAccumulatedStats calls reset_peak() and resetPeakStats calls reset_accumulated(). Can you double-check that?
> 

Good catch on this, I've fixed it.

> Thread safety on stats_ :recordAlloc and recordRelease mutate the static stats_ member on every allocation/deallocation without any visible locking. If allocations can happen concurrently from different threads, this could race. Worth considering whether a mutex or atomics are needed here.

We don't have multi-thread in spyre as of now, @ppnaik1890 and I will open an issue and implement this in a follow-up PR.

> 
> Test coverage for the edge case: it would be good to add a test that calls memory_allocated() before any Spyre tensor has been created, since that's the exact scenario driving the guard discussion.

Previous tests in that file will initialize tensors, so unfortunately we will not be able to test this exact scenario.

**Review by thoangtrvn — APPROVED** [review]
LGTM. 
The code now adopts `torch.accelerator.memory.<API>` properly.  We can use both also with `torch.spyre.memory.<API>` which always mirrors API from Pytorch upstream. It also works with lazy loading mechanism.

**Review by thoangtrvn — CHANGES_REQUESTED** [review]
Based on the comment from @kmehant , I agree that we should limit the scope of mask in the monkey-patch to the exact error. I have included a proposed code fix.

**Review by thoangtrvn — APPROVED** [review]

**Review by kmehant — APPROVED** [review]

**Review by avery-blanchard — APPROVED** [review]
